projects/models.nix

```diff
@@ -0,0 +1,152 @@
+{
+  lib,
+  pkgs,
+  sources,
+  yants ? import sources.yants { inherit lib; },
+}:
+let
+  inherit (yants)
+    string
+    list
+    option
+    attrs
+    either
+    struct
+    drv
+    path
+    restrict
+    defun
+    function
+    any
+    ;
+
+  mkModule =
+    defun
+      [
+        (attrs any)
+        moduleType
+      ]
+      (
+        { name, src, ... }:
+        {
+          name = "hello ${name}";
+          inherit src;
+        }
+      );
+
+  # WIP: just an example for now
+  moduleType = struct {
+    name = string;
+    src = string;
+  };
+
+  programType = struct "program" {
+    name = option string;
+    module = moduleType;
+    documentation = optionalStruct {
+      build = option string;
+      tests = option string;
+    };
+    examples = nonEmtpyAttrs (option exampleType);
+  };
+
+  serviceType = struct "service" {
+    name = option string;
+    path = either string path;
+    documentation = optionalStruct {
+      config = option string;
+    };
+    examples = nonEmtpyAttrs (option exampleType);
+  };
+
+  exampleType = struct "example" {
+    description = option string; # TODO: should this be non-optional?
+    path = either string path;
+    documentation = option string;
```

Comment:
Generally for reference we probably want a structured URL type along these lines:

```nix
  urlType = struct "URL" {
    text = string;
    # we may later want to do a fancy syntax check in a custom `typdef`
    link = string;
  };
```

Reply:
This makes sense to me. I assume we should do the same for `programs` and `services`, but do we want to make them into an `attrs` or restrict which types of links are allowed with `struct`?

> ... and possibly later auto-add a __toString attribute or even various output formats so we can string interpolate this in e.g. example descriptions for correct-by-construction URLs.

Yup, I believe this is possible as well.

Reply:
Where we know (or expect) a certain set of URLs I'd go for struct. Free form stuff such as additinal references can be attrs.